File: app/src/main/java/com/willchou/dapenti/presenter/PageOrderActivity.java
package com.willchou.dapenti.presenter;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.Toolbar;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.TextView;

import com.willchou.dapenti.R;
import com.willchou.dapenti.model.DaPenTi;
import com.willchou.dapenti.model.Database;
import com.woxthebox.draglistview.DragItem;
import com.woxthebox.draglistview.DragItemAdapter;
import com.woxthebox.draglistview.DragListView;

import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class PageOrderActivity extends AppCompatActivity {
    private static final String TAG = "PageOrderActivity";

    private Button applyButton;
    private DragListView dragListView;

    class ItemDetail { String title; long uniqueID; boolean visible; }
    private List<ItemDetail> itemList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_page_order);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener((View v) -> onBackPressed());

        setupData();
        setupLayout();
    }

    private void setupData() {
        Database database = Database.getDatabase();

        if (database == null)
            return;

        List<Pair<String, URL>> categories = new ArrayList<>();
        database.getCategories(categories, false);
        List<Pair<String, Boolean>> visibleList = new ArrayList<>();
        database.getCategoryVisible(visibleList);

        int count = 0;
        for (Pair<String, URL> p : categories) {
            ItemDetail id = new ItemDetail();
            id.title = p.first; id.uniqueID = count ++;

            for (Pair<String, Boolean> p2 : visibleList)
                if (p2.first.equals(id.title)) {
                    id.visible = p2.second;
                    break;
                }

            itemList.add(id);
        }
    }

    private void setupLayout() {
        applyButton = findViewById(R.id.apply_button);
        applyButton.setVisibility(View.GONE);
        applyButton.setOnClickListener(view -> saveChanged());

        dragListView = findViewById(R.id.drag_list_view);
        dragListView.getRecyclerView().setVerticalScrollBarEnabled(true);
        dragListView.setLayoutManager(new LinearLayoutManager(this));
        dragListView.setAdapter(new ItemAdapter(), true);
        dragListView.setCanDragHorizontally(false);
        dragListView.setCustomDragItem(new MyDragItem(this, R.layout.page_list_item));
        dragListView.setDragListListener(new DragListView.DragListListenerAdapter() {
            @Override
            public void onItemDragEnded(int fromPosition, int toPosition) {
                if (fromPosition != toPosition && applyButton.getVisibility() != View.VISIBLE)
                    applyButton.setVisibility(View.VISIBLE);
            }
        });
    }

    private void saveChanged() {
        applyButton.setVisibility(View.GONE);

        List<Pair<String, Boolean>> list = new ArrayList<>();
        for (ItemDetail id : itemList)
            list.add(new Pair<>(id.title, id.visible));

        if (Database.getDatabase() != null) {
            Database.getDatabase().updateCategoriesOrderAndVisible(list);
            if (DaPenTi.getDaPenTi() != null)
                DaPenTi.getDaPenTi().prepareCategory(false);
            Snackbar.make(dragListView, "应用成功", Snackbar.LENGTH_SHORT).show();
        } else {
            Snackbar.make(dragListView, "数据库操作失败", Snackbar.LENGTH_SHORT)
                    .show();
            applyButton.setVisibility(View.VISIBLE);
        }
    }

    class ItemAdapter extends DragItemAdapter<ItemDetail, PageOrderActivity.ViewHolder> {
        ItemAdapter() {
            setItemList(itemList);
        }

        @Override
        public void onBindViewHolder(@NonNull PageOrderActivity.ViewHolder holder, int position) {
            super.onBindViewHolder(holder, position);

            ItemDetail id = itemList.get(position);

            holder.itemDetail = id;
            holder.textView.setText(id.title);
            holder.checkBox.setChecked(id.visible);
        }

        @NonNull
        @Override
        public PageOrderActivity.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(getBaseContext()).inflate(R.layout.page_list_item, parent, false);
            return new PageOrderActivity.ViewHolder(view, R.id.drag_image, false);
        }

        @Override
        public long getUniqueItemId(int position) {
            return itemList.get(position).uniqueID;
        }
    }

    class ViewHolder extends DragItemAdapter.ViewHolder {
        TextView textView;
        CheckBox checkBox;
        ItemDetail itemDetail;

        ViewHolder(View itemView, int handleResId, boolean dragOnLongPress) {
            super(itemView, handleResId, dragOnLongPress);
            textView = itemView.findViewById(R.id.drag_page_name);
            checkBox = itemView.findViewById(R.id.visible_checkbox);

            checkBox.setOnCheckedChangeListener((compoundButton, b) -> new Exception().printStackTrace());
        }

        @Override
        public void onItemClicked(View view) {
            super.onItemClicked(view);
            checkBox.toggle();

            itemDetail.visible = checkBox.isChecked();

            if (applyButton.getVisibility() != View.VISIBLE)
                applyButton.setVisibility(View.VISIBLE);
        }
    }

    class MyDragItem extends DragItem {
        MyDragItem(Context context, int layoutId) {
            super(context, layoutId);
        }

        @Override
        public void onBindDragView(View clickedView, View dragView) {
            CharSequence text = ((TextView) clickedView.findViewById(R.id.drag_page_name)).getText();
            ((TextView) dragView.findViewById(R.id.drag_page_name)).setText(text);

            boolean checked = ((CheckBox) clickedView.findViewById(R.id.visible_checkbox)).isChecked();
            ((CheckBox) (dragView.findViewById(R.id.visible_checkbox))).setChecked(checked);
        }
    }
}

File: app/src/main/java/com/willchou/dapenti/presenter/PageOrderActivity.kt
package com.willchou.dapenti.presenter

import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.CardView
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.CheckBox
import android.widget.TextView
import com.willchou.dapenti.R
import com.willchou.dapenti.model.Settings
import com.willchou.dapenti.vm.OrderViewModel
import com.woxthebox.draglistview.DragItem
import com.woxthebox.draglistview.DragItemAdapter
import com.woxthebox.draglistview.DragListView
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import me.majiajie.swipeback.SwipeBackActivity
import kotlin.collections.ArrayList

class PageOrderActivity : SwipeBackActivity() {
    companion object {
        private const val TAG = "PageOrderActivity"
    }

    private var applyButton: Button? = null
    private var dragListView: DragListView? = null

    private var orderViewModel: OrderViewModel? = null

    private val itemDetailList = ArrayList<ItemDetail>()

    internal inner class ItemDetail (
            var title: String,
            var uniqueID: Long,
            var visible: Boolean)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_page_order)

        val toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        toolbar.setNavigationOnClickListener { onBackPressed() }

        if (Settings.settings!!.nightMode) {
            setTheme(R.style.NightModeTheme)

            val color = Settings.settings!!.getBackgroundColor()
            findViewById(R.id.drag_list_view)?.setBackgroundColor(color)
            findViewById(R.id.linearLayout)?.setBackgroundColor(color)
        }

        Observable.just(this)
                .subscribeOn(Schedulers.io())
                .doOnNext {
                    orderViewModel = ViewModelProviders.of(this).get(OrderViewModel::class.java)

                    val list = orderViewModel!!.getCategoryAndVisibility()
                    for ((i, l) in list.withIndex())
                        itemDetailList.add(ItemDetail(l.first, i.toLong(), l.second))
                }
                .subscribe { runOnUiThread { setupLayout() } }
    }

    private fun setupLayout() {
        applyButton = findViewById(R.id.apply_button) as Button
        applyButton?.visibility = View.GONE
        applyButton?.setOnClickListener { saveChanged() }

        dragListView = findViewById(R.id.drag_list_view) as DragListView
        if (dragListView == null)
            return

        dragListView?.recyclerView?.isVerticalScrollBarEnabled = true
        dragListView?.setLayoutManager(LinearLayoutManager(this))
        dragListView?.setAdapter(ItemAdapter(), true)
        dragListView?.setCanDragHorizontally(false)
        dragListView?.setCustomDragItem(MyDragItem(this, R.layout.page_list_item))
        dragListView?.setDragListListener(object : DragListView.DragListListenerAdapter() {
            override fun onItemDragEnded(fromPosition: Int, toPosition: Int) {
                if (fromPosition != toPosition && applyButton!!.visibility != View.VISIBLE)
                    applyButton!!.visibility = View.VISIBLE
            }
        })
    }

    private fun saveChanged() {
        applyButton?.visibility = View.GONE

        val list: MutableList<Pair<String, Boolean>> = ArrayList()
        for (id in itemDetailList)
            list.add(Pair(id.title, id.visible))

        Observable.just(list)
                .subscribeOn(Schedulers.io())
                .doOnNext { orderViewModel!!.saveCategoryOrderAndVisibility(list) }
                .subscribe {
                    runOnUiThread {
                        Snackbar.make(findViewById(android.R.id.content), "操作成功", Snackbar.LENGTH_SHORT)
                                .show()
                        applyButton!!.visibility = View.GONE
                    }
                }
    }

    internal inner class ItemAdapter : DragItemAdapter<ItemDetail, ViewHolder>() {
        init {
            itemList = itemDetailList
        }

        override fun onBindViewHolder(holder: PageOrderActivity.ViewHolder, position: Int) {
            super.onBindViewHolder(holder, position)

            val id = itemList[position]

            holder.itemDetail = id
            holder.textView.text = id.title
            holder.checkBox.isChecked = id.visible
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int):
                PageOrderActivity.ViewHolder {
            val view = LayoutInflater
                    .from(baseContext)
                    .inflate(R.layout.page_list_item, parent, false)

            if (Settings.settings!!.nightMode) {
                view.findViewById<CardView>(R.id.cardView)
                        ?.setBackgroundColor(Settings.settings!!.getLighterBackgroundColor())
                view.setBackgroundColor(Settings.settings!!.getBackgroundColor())
            }

            return ViewHolder(view, R.id.dragImage, false, applyButton)
        }

        override fun getUniqueItemId(position: Int): Long {
            return itemList[position].uniqueID
        }
    }

    internal inner class ViewHolder(itemView: View, handleResId: Int, dragOnLongPress: Boolean,
                                    private val triggerButton: View?) :
            DragItemAdapter.ViewHolder(itemView, handleResId, dragOnLongPress) {
        var textView: TextView = itemView.findViewById(R.id.pageName)
        var checkBox: CheckBox = itemView.findViewById(R.id.visibleCheckbox)
        var itemDetail: ItemDetail? = null

        override fun onItemClicked(view: View?) {
            super.onItemClicked(view)
            checkBox.toggle()
            itemDetail?.visible = checkBox.isChecked

            if (triggerButton?.visibility != View.VISIBLE)
                triggerButton?.visibility = View.VISIBLE
        }
    }

    internal inner class MyDragItem(context: Context, layoutId: Int) : DragItem(context, layoutId) {
        override fun onBindDragView(clickedView: View, dragView: View) {
            val text = clickedView.findViewById<TextView>(R.id.pageName).text
            val checked = clickedView.findViewById<CheckBox>(R.id.visibleCheckbox).isChecked

            (dragView.findViewById<View>(R.id.pageName) as TextView).text = text
            (dragView.findViewById<View>(R.id.visibleCheckbox) as CheckBox).isChecked = checked

            if (Settings.settings!!.nightMode)
                dragView.setBackgroundColor(Settings.settings!!.getBackgroundColor())
        }
    }
}
